pkg/constraints/constraint.go

```diff
@@ -0,0 +1,83 @@
+package constraints
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+)
+
+// OLMConstraintType is the schema "type" key for all constraints known to OLM
+// (except for legacy types).
+const OLMConstraintType = "olm.constraint"
+
+// Constraint holds parsed, potentially nested dependency constraints.
+type Constraint struct {
+	// Constraint message that surfaces in resolution
+	// This field is optional
+	Message string `json:"message"`
+
+	// The cel struct that contraints CEL expression
+	Cel *Cel `json:"cel,omitempty" yaml:"cel,omitempty"`
+
+	// Package defines a constraint for a package within a version range.
+	Package *PackageConstraint `json:"package,omitempty" yaml:"package,omitempty"`
+	// GVK defines a constraint for a GVK.
+	GVK *GVKConstraint `json:"gvk,omitempty" yaml:"gvk,omitempty"`
+
+	// All, Any, and None are compound constraints. See this enhancement for details:
+	// https://github.com/operator-framework/enhancements/blob/master/enhancements/compound-bundle-constraints.md
+	All *CompoundConstraint `json:"all,omitempty" yaml:"all,omitempty"`
+	Any *CompoundConstraint `json:"any,omitempty" yaml:"any,omitempty"`
+	// A note on None: this constraint is not particularly useful by itself.
+	// It should be used within an All constraint alongside some other constraint type
+	// since saying "none of these GVKs/packages/etc." without an alternative doesn't make sense.
+	None *CompoundConstraint `json:"none,omitempty" yaml:"none,omitempty"`
+}
+
+// CompoundConstraint holds a list of potentially nested constraints
+// over which a boolean operation is applied.
+type CompoundConstraint struct {
+	Constraints []Constraint `json:"constraints"`
```

Comment:
While I understand `CompoundConstraint` is a list of constraints but then `Constraint` itself can be a `All`, `Any`, `None` constraint which is a `CompoundConstraint` itself. So we potentially have a `CompoundConstraint` as a list of `CompoundConstraint`. While it is not a problem programmatically but logically it is convoluted and frankly difficult to test.

Reply:
Not quite, since `CompoundConstraint` does not alias `Constraint`; they're two distinct types. I don't see how you would get nice recursion that leverages std serialization libs otherwise. The [encoded form](https://github.com/operator-framework/operator-lifecycle-manager/pull/2418/files#diff-ea16582ca96e853d17883479654c218897cb4bf126ddd3c0d4908d24ef99329eR806-R812) is readable and comprehensible imo. Plus this is what the EP proposed.